### PR TITLE
Fixed Path of Exile not rendering correctly.

### PR DESCRIPTION
**Pull request by AnzhongHuang**
The regular sequence of translation of exp2 should be: @llvm.exp2.f32 ->
@ldexpf -> @llvm.amdgcn.ldexp.f32. But in this case, the @ldexpf
instruction will not be translated by Opt phase, and then it's passed to
back-end, but this instruction is not recognized by LLPC back-end, so it
will be dropped.

This change adds an extra Link phase after Opt phase, it can resolve the
@ldexpf after Opt phase.

### CODE REVIEW

File: patch/generate/llvmIntrinsicEmu.ll [hunk 1]
```diff
@@ -29,6 +29,15 @@ define float @tanf(float %angle) #0
     ret float %1
 }
 
+; GLSL: float ldexpf(float, int)
+define float @ldexpf(float %x, i32 %exp)
+{
+    ; This function is generated by LLVM call simplification when handling pow(2.0, exp).
```
Comment (linqun) [inline patch/generate/llvmIntrinsicEmu.ll]: please remove when handling pow(2.0, exp). in comments, it doesn't match your case

File: patch/generate/llvmIntrinsicEmu.ll [hunk 2]
```diff
@@ -29,6 +29,15 @@ define float @tanf(float %angle) #0
     ret float %1
 }
 
+; GLSL: float ldexpf(float, int)
```
Comment (linqun) [inline patch/generate/llvmIntrinsicEmu.ll]: remove GLSL: it is not from GLSL built-in function.

File: patch/generate/llvmIntrinsicEmu.ll [hunk 3]
```diff
@@ -29,6 +29,15 @@ define float @tanf(float %angle) #0
...
+define float @ldexpf(float %x, i32 %exp)
+{
+    ; This function is generated by LLVM call simplification when handling pow(2.0, exp).
+    %1 = call float @llvm.amdgcn.ldexp.f32(float %x, i32 %exp) #1
```
Comment (linqun) [inline patch/generate/llvmIntrinsicEmu.ll]: #1 => #0

File: patch/generate/llvmIntrinsicEmu.ll [hunk 4]
```diff
@@ -29,6 +29,15 @@ define float @tanf(float %angle) #0
...
+    ret float %1
+}
+
 declare float @llpc.tan.f32(float %angle) #0
+declare float @llvm.amdgcn.ldexp.f32(float, i32) #1
```
Comment (linqun) [inline patch/generate/llvmIntrinsicEmu.ll]: #1 => #0

File: patch/llpcPatch.cpp [hunk 5]
```diff
@@ -155,6 +155,10 @@ void Patch::AddPasses(
     if (cl::DisablePatchOpt == false)
     {
         AddOptimizationPasses(pContext, passMgr);
+
+        // Link external libraries and remove dead functions after it
+        passMgr.add(CreatePassExternalLibLink(false)); // Not native only
```
Comment (linqun) [inline patch/llpcPatch.cpp]: remove  // Not native only

File: patch/llpcPatch.cpp [hunk 6]
```diff
@@ -155,6 +155,10 @@ void Patch::AddPasses(
     if (cl::DisablePatchOpt == false)
     {
         AddOptimizationPasses(pContext, passMgr);
+
+        // Link external libraries and remove dead functions after it
```
Comment (linqun) [inline patch/llpcPatch.cpp]: refine comments:  Link external libraries to solve LLVM library func ...

**Comment by amdvlk-admin** [comment]
### Test summary for commit 66178f32219286031f76a2b36a0da5659eea24e1
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev e8b82e5f183f4cfcf4f7b893295c49d0fc929fa0</li>
<li>SPVGEN: dev 1560d287f779b342e0019499dda85890cb07affc</li>
<li>XGL: dev b5440b6a385dbcb11739f1bdb89a94bbe5cd499f</li>
<li>LLVM: amd-vulkan-dev 97cc33415120ae3ed472b6dd5cb234b74a80bd80</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 902/902 (100.0%)</li>
<li>Failed: 0/902 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit ef4687ae6b1c13e89a3682050639f5704be85bf2
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev e8b82e5f183f4cfcf4f7b893295c49d0fc929fa0</li>
<li>SPVGEN: dev 1560d287f779b342e0019499dda85890cb07affc</li>
<li>XGL: dev b5440b6a385dbcb11739f1bdb89a94bbe5cd499f</li>
<li>LLVM: amd-vulkan-dev 97cc33415120ae3ed472b6dd5cb234b74a80bd80</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 902/902 (100.0%)</li>
<li>Failed: 0/902 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit bbb419494083f953521f2fc61187f241356d2f27
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev e8b82e5f183f4cfcf4f7b893295c49d0fc929fa0</li>
<li>SPVGEN: dev 1560d287f779b342e0019499dda85890cb07affc</li>
<li>XGL: dev b5440b6a385dbcb11739f1bdb89a94bbe5cd499f</li>
<li>LLVM: amd-vulkan-dev 97cc33415120ae3ed472b6dd5cb234b74a80bd80</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 902/902 (100.0%)</li>
<li>Failed: 0/902 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.1.2-cherry-picks)</summary><ul>


<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 312622/312622 (100.0%)</li>
<li>Failed: 0/312622 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**